clippy_lints/src/needless_traits_in_scope.rs

```diff
@@ -0,0 +1,86 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Find traits that are not explicitely used in scope (like `AsRef::as_ref()`)
+    /// but directly with the trait method (like `opt.as_ref()`).
+    ///
+    /// These traits can be imported anonymously with `use crate::Trait as _`.
+    /// This avoids name collision with other traits (possibly with the same name).
+    /// It also helps identify the traits in `use` statements.
+    ///
+    /// ### Why is this bad?
+    /// This needlessly brings a trait in trait's namespace, where it could
```

Comment:
Trait names do not have their own namespace; rather, [they live in the type namespace](https://doc.rust-lang.org/reference/names/namespaces.html).
```suggestion
    /// This needlessly brings a trait into the type namespace, where it could
```

This could be another reason for someone to use this trait: it'd allow them to avoid name collisions between types and traits.

Reply:
Indeed, thank you, I had actually problem to find the information (probably missing the right keywords). I remember there were a few different namespaces, but couldn't remember the exact boundaries of each category.

Fixed in a6e854e62edbabcb33f8f10658c9e098e23d4868.